tstest: ResourceCheck swallows panics

Due to https://github.com/golang/go/issues/49929, [`tstest.ResourceCheck`](https://pkg.go.dev/tailscale.com/tstest#ResourceCheck) ends up swallowing panics in functions where the close/cancel funcs don't get called due to the panic. For example:

```go
func TestResourceCheck(t *testing.T) {
	tstest.ResourceCheck(t)

	// Start a goroutine and return a close function, which never gets called
	// due to a panic in the function.
	close := func() context.CancelFunc {
		ctx, cancel := context.WithCancel(context.Background())
		go func() { <-ctx.Done() }()
		panic("uhoh")
		return cancel
	}
	defer close()
}
```

Produces the following output instead of the "panic: uhoh" and stack trace.

```
--- FAIL: TestResourceCheck (3.29s)
    /Users/icio/go/src/github.com/tailscale/testres/resource.go:46: goroutine diff:
          bytes.Join({
        - 	"goroutine profile: total 2",
        - 	"1 @ 0x100339f0c 0x10037b8b4 0x1004f6af4 0x1004f6910 0x1004f3ea8 0x10140d564 0x10140d17c 0x1014c17a0 0x10044cf14 0x100385ba4",
        - 	`#	0x1004f6af3	runtime/pprof.writeRuntimeProfile+0xb3		/Users/icio/.cache/tailscale-go/src/runtime/pprof/pprof.go:793`,
        - 	`#	0x1004f690f	runtime/pprof.writeGoroutine+0x4f		/Users/icio/.cache/tailscale-go/src/runtime/pprof/pprof.go:752`,
        - 	`#	0x1004f3ea7	runtime/pprof.(*Profile).WriteTo+0x147		/Users/icio/.cache/tailscale-go/src/runtime/pprof/pprof.go:374`,
        - 	`#	0x10140d563	tailscale.com/tstest.goroutines+0x53		/Users/icio/go/pkg/mod/tailscale.com@v1.75.0-pre.0.20241010161436-508980603b39/tstest/resource.go:54`,
        - 	`#	0x10140d17b	tailscale.com/tstest.ResourceCheck+0x5b		/Users/icio/go/pkg/mod/tailscale.com@v1.75.0-pre.0.20241010161436-508980603b39/tstest/resource.go:29`,
        - 	`#	0x1014c179f	restest.TestResourceCheck+0x2f	/Users/icio/go/src/github.com/tailscale/testres/res_test.go:4336`,
        - 	`#	0x10044cf13	testing.tRunner+0xe3				/Users/icio/.cache/tailscale-go/src/testing/testing.go:1690`,
        + 	"goroutine profile: total 3",
        + 	"1 @ 0x100339f0c 0x10037b8b4 0x1004f6af4 0x1004f6910 0x1004f3ea8 0x10140d564 0x10140d340 0x10044bb18 0x10044c500 0x10044cfbc 0x10037c704 0x1014f3170 0x1014c17d0 0x10044cf14 0x100385ba4",
        + 	`#	0x1004f6af3	runtime/pprof.writeRuntimeProfile+0xb3			/Users/icio/.cache/tailscale-go/src/runtime/pprof/pprof.go:793`,
        + 	`#	0x1004f690f	runtime/pprof.writeGoroutine+0x4f			/Users/icio/.cache/tailscale-go/src/runtime/pprof/pprof.go:752`,
        + 	`#	0x1004f3ea7	runtime/pprof.(*Profile).WriteTo+0x147			/Users/icio/.cache/tailscale-go/src/runtime/pprof/pprof.go:374`,
        + 	`#	0x10140d563	tailscale.com/tstest.goroutines+0x53			/Users/icio/go/pkg/mod/tailscale.com@v1.75.0-pre.0.20241010161436-508980603b39/tstest/resource.go:54`,
        + 	`#	0x10140d33f	tailscale.com/tstest.ResourceCheck.func1+0xef		/Users/icio/go/pkg/mod/tailscale.com@v1.75.0-pre.0.20241010161436-508980603b39/tstest/resource.go:42`,
        + 	`#	0x10044bb17	testing.(*common).Cleanup.func1+0xf7			/Users/icio/.cache/tailscale-go/src/testing/testing.go:1176`,
        + 	`#	0x10044c4ff	testing.(*common).runCleanup+0xcf			/Users/icio/.cache/tailscale-go/src/testing/testing.go:1354`,
        + 	`#	0x10044cfbb	testing.tRunner.func2+0x2b				/Users/icio/.cache/tailscale-go/src/testing/testing.go:1684`,
        + 	`#	0x10037c703	runtime.gopanic+0x123					/Users/icio/.cache/tailscale-go/src/runtime/panic.go:785`,
        + 	`#	0x1014f316f	restest.TestResourceCheck.func1+0x7f	/Users/icio/go/src/github.com/tailscale/testres/res_test.go:4345`,
        + 	`#	0x1014c17cf	restest.TestResourceCheck+0x5f		/Users/icio/go/src/github.com/tailscale/testres/res_test.go:4349`,
        + 	`#	0x10044cf13	testing.tRunner+0xe3					/Users/icio/.cache/tailscale-go/src/testing/testing.go:1690`,
          	"",
          	"1 @ 0x10037caa8 0x10030fc44 0x10030f7f4 0x10044dc28 0x10044fc30 0x10044cf14 0x10044fb2c 0x10044e878 0x1015b18b0 0x100344a38 0x100385ba4",
          	... // 5 identical lines
          	`#	0x1015b18af	main.main+0x8f			_testmain.go:239`,
          	`#	0x100344a37	runtime.main+0x287		/Users/icio/.cache/tailscale-go/src/runtime/proc.go:272`,
        + 	"",
        + 	"1 @ 0x10037caa8 0x10030fc44 0x10030f7f4 0x1014f31c0 0x100385ba4",
        + 	`#	0x1014f31bf	restest.TestResourceCheck.func1.1+0x2f	/Users/icio/go/src/github.com/tailscale/testres/res_test.go:4344`,
          	"",
          	"",
          }, "\n")
```

--- a/tstest/resource.go
+++ b/tstest/resource.go
@@ -29,7 +29,8 @@ func ResourceCheck(tb testing.TB) {
 	startN, startStacks := goroutines()
 	tb.Cleanup(func() {
 		if tb.Failed() {
-			// Something else went wrong.
+			// Test has failed - but this doesn't catch panics due to
+			// https://github.com/golang/go/issues/49929.
 			return
 		}
 		// Goroutines might be still exiting.
@@ -44,7 +45,10 @@ func ResourceCheck(tb testing.TB) {
 			return
 		}
 		tb.Logf("goroutine diff:\n%v\n", cmp.Diff(startStacks, endStacks))
-		tb.Fatalf("goroutine count: expected %d, got %d\n", startN, endN)
+
+		// tb.Failed() above won't report on panics, so we shouldn't call Fatal
+		// here or we risk suppressing reporting of the panic.
+		tb.Errorf("goroutine count: expected %d, got %d\n", startN, endN)
 	})
 }
 

